File: src/main/java/pl/yalgrin/playnite/simplesync/domain/objects/Game.java
package pl.yalgrin.playnite.simplesync.domain.objects;

import io.r2dbc.postgresql.codec.Json;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.Transient;
import org.springframework.data.relational.core.mapping.Column;
import org.springframework.data.relational.core.mapping.Table;

@EqualsAndHashCode(callSuper = true)
@Data
@NoArgsConstructor
@AllArgsConstructor
@SuperBuilder
@Table("playnite_game")
public class Game extends AbstractObjectEntity {
    @Id
    @Column("id")
    private Long id;

    @Column("playnite_id")
    private String playniteId;

    @Column("game_id")
    private String gameId;

    @Column("plugin_id")
    private String pluginId;

    @Column("name")
    private String name;

    @Column("contents")
    private Json savedData;

    @Column("icon_md5")
    private String iconMd5;

    @Column("cover_image_md5")
    private String coverImageMd5;

    @Column("background_image_md5")
    private String backgroundImageMd5;

    @Column("removed")
    private boolean removed;

    @Transient
    private boolean notifyAll;

    @Transient
    private boolean changed;
}

File: src/main/kotlin/pl/yalgrin/playnite/simplesync/library/domain/Game.kt
package pl.yalgrin.playnite.simplesync.library.domain

import io.r2dbc.postgresql.codec.Json
import org.springframework.data.annotation.Id
import org.springframework.data.annotation.Transient
import org.springframework.data.relational.core.mapping.Column
import org.springframework.data.relational.core.mapping.Table
import pl.yalgrin.playnite.simplesync.common.util.asObject
import pl.yalgrin.playnite.simplesync.library.dto.GameDatabaseModel
import reactor.core.publisher.Mono
import java.time.Instant

object GameConstants {
    const val CURRENT_MODEL_VERSION = 3L
}

@Table("playnite_game")
data class Game(
    @Id
    @Column("id")
    override var id: Long? = null,

    @Column("playnite_id")
    override var playniteId: String? = null,

    @Column("game_id")
    var gameId: String? = null,

    @Column("plugin_id")
    var pluginId: String? = null,

    @Column("name")
    override var name: String? = null,

    @Column("contents")
    var savedData: Json? = null,

    @Column("model_version")
    var modelVersion: Long = GameConstants.CURRENT_MODEL_VERSION,

    @Column("icon_md5")
    var iconMd5: String? = null,

    @Column("cover_image_md5")
    var coverImageMd5: String? = null,

    @Column("background_image_md5")
    var backgroundImageMd5: String? = null,

    @Column("removed")
    override var isRemoved: Boolean = false,

    @Column("created_at")
    override var createdAt: Instant = Instant.now(),

    @Column("created_by")
    override var createdBy: String? = null,

    @Column("modified_at")
    override var modifiedAt: Instant = Instant.now(),

    @Column("modified_by")
    override var modifiedBy: String? = null,

    @Transient
    override var isNotifyAll: Boolean = false,

    @Transient
    override var isChanged: Boolean = false
) : LibraryObjectEntity

fun Game.extractDbModelOrEmpty(): Mono<GameDatabaseModel> {
    return this.savedData.asObject(GameDatabaseModel::class.java)
        .switchIfEmpty(Mono.fromSupplier { GameDatabaseModel() })
}